modules/noise/noise.h

```diff
@@ -0,0 +1,239 @@
+			(T *)p_src->get_data().ptr(),
+			src_width, src_height,
+			half_width, half_height,
+			p_width, p_height
+		};
+
+		// `wr` is setup for straight x/y coordinate array access.
+		img_buff<T> wr = {
+			(T *)dest.ptrw(),
+			p_width, p_height,
+			0, 0, 0, 0
+		};
+		// `rd_dest` is a readable pointer to `wr`, i.e. what has already been written to the output buffer.
+		img_buff<T> rd_dest = {
+			(T *)dest.ptr(),
+			p_width, p_height,
+			0, 0, 0, 0
+		};
+
+		// Swap the quadrants to make edges seamless.
+		for (int y = 0; y < p_height; y++) {
+			for (int x = 0; x < p_width; x++) {
+				// rd_src has a half offset and the shorter modulo ignores the skirt.
+				// It reads and writes in Q1-4 order (see map above), skipping the skirt.
+				wr(x, y) = rd_src(x, y, img_buff<T>::ALT_XY);
+			}
+		}
+
+		// Blend the vertical skirt over the middle seam.
+		for (int x = half_width; x < skirt_edge_x; x++) {
+			int alpha = 255 * (1 - Math::smoothstep(.1f, .9f, float(x - half_width) / float(skirt_width)));
+			for (int y = 0; y < p_height; y++) {
+				// Skip the center square
+				if (y == half_height) {
+					y = skirt_edge_y - 1;
+				} else {
+					// Starts reading at s2, ALT_Y skips s3, and continues with s1.
+					wr(x, y) = _alpha_blend<T>(rd_dest(x, y), rd_src(x, y, img_buff<T>::ALT_Y), alpha);
+				}
+			}
+		}
+
+		// Blend the horizontal skirt over the middle seam.
+		for (int y = half_height; y < skirt_edge_y; y++) {
+			int alpha = 255 * (1 - Math::smoothstep(.1f, .9f, float(y - half_height) / float(skirt_height)));
+			for (int x = 0; x < p_width; x++) {
+				// Skip the center square
+				if (x == half_width) {
+					x = skirt_edge_x - 1;
+				} else {
+					// Starts reading at s4, skips s3, continues with s5.
+					wr(x, y) = _alpha_blend<T>(rd_dest(x, y), rd_src(x, y, img_buff<T>::ALT_X), alpha);
+				}
+			}
+		}
+
+		// Fill in the center square. Wr starts at the top left of Q4, which is the equivalent of the top left of s3, unless a modulo is used.
+		for (int y = half_height; y < skirt_edge_y; y++) {
+			for (int x = half_width; x < skirt_edge_x; x++) {
+				int xpos = 255 * (1 - Math::smoothstep(.1f, .9f, float(x - half_width) / float(skirt_width)));
+				int ypos = 255 * (1 - Math::smoothstep(.1f, .9f, float(y - half_height) / float(skirt_height)));
+
+				// Blend s3(Q1) onto s5(Q2) for the top half.
+				T top_blend = _alpha_blend<T>(rd_src(x, y, img_buff<T>::ALT_X), rd_src(x, y, img_buff<T>::DEFAULT), xpos);
+				// Blend s1(Q3) onto Q4 for the bottom half.
+				T bottom_blend = _alpha_blend<T>(rd_src(x, y, img_buff<T>::ALT_XY), rd_src(x, y, img_buff<T>::ALT_Y), xpos);
+				// Blend the top half onto the bottom half.
+				wr(x, y) = _alpha_blend<T>(bottom_blend, top_blend, ypos);
+			}
+		}
+		Ref<Image> image = memnew(Image(p_width, p_height, false, p_src->get_format(), dest));
+		p_src.unref();
+		return image;
+	}
+
+	template <typename T>
+	T _alpha_blend(T p_bg, T p_fg, int p_alpha) const {
```

Comment:
What was the idea behind converting this and all of these functions to a template? What other types will be used here? This alpha_blend function is specifically designed for a 32-bit long int, the `l` component of the l2c union. I think there will be unexpected results with other types.

Are there any active use cases right now for generating non-32-bit noise images? Were they tested and do they work, including with the alpha blend seamless function?

Reply:
I think the purpose of this was to allow for grayscale image optimizations when a noise generator creates an image with `Image::FORMAT_L8` (see the template specialization and the get_seamless_image implementation in noise.cpp). 
But looking at the code now I think there is room for improvement, which I might include in a subsequent PR. E.g. I can't remember why I didn't just overload the alpha blend function.